content/articles/2018-11-tibble-2.0.0-announce.Rmd

```diff
@@ -0,0 +1,158 @@
+Please be aware that `nrow` must be passed as named argument, because it comes after the ellipsis `...` in the signature.
+
+```{r error = TRUE}
+x <- data.frame(a = 1)
+
+# Old:
+new_tibble(x)
+
+# New:
+new_tibble(x, nrow = nrow(x)) # if x is a data frame
+
+nrow_x <- NROW(x[[1]]) # if x has at least one column
+# nrow_x <- ... # if the number of rows is given elsewhere
+new_tibble(x, nrow = nrow_x)
+```
+
+## Coercion and name repair
+
+There are some rare legitimate scenarios involving tibbles with empty or duplicate column names.
+For instance, after importing data from a foreign format, only inspection of the data might reveal if the column is useful in the first place.
+In the vast majority of cases, empty and duplicate names should be avoided, ideally the names should be accessible in R without using backticks.
+The `as_tibble()` method and the `tibble()` constructor now supports a new `.name_repair` argument that covers most use cases:
+
+- `"minimal"`: No name repair or checks, beyond basic existence.
+- `"unique"`: Make sure names are unique and not empty.
+- `"check_unique"`: (default value), no name repair, but check they are `unique`.
+- `"universal"`: Make the names `unique` and syntactic.
+- a function: apply custom name repair (e.g., `.name_repair = make.names` or `.name_repair = ~make.names(., unique = TRUE)` for names in the style of base R).
+
+Name repair is applied always, even when coercing from a matrix or a tibble.
+It is now harder (though not impossible) to construct a tibble with empty or duplicate column names.
+See `?repair_names` for details on the new name repair algorithms.
+
+```{r error = TRUE}
+m <- cov(unname(iris[-5]))
+
+# Old:
+df <- as_tibble(m)
+colnames(df) <- letters[1:4]
+
+# New (preferred):
+colnames(m) <- letters[1:4]
+df <- as_tibble(m)
+
+# New (alternative):
+df <- as_tibble(m, .name_repair = "minimal")
+colnames(df) <- letters[1:4]
+```
+
+The `validate` argument is deprecated and gives a message.
+If your code must support _tibble_ < 2.0.0, please use the new code only if `packageVersion("tibble") >= "2.0.0"` .
+
+```{r error = TRUE}
+m <- cov(unname(iris[-5]))
+df <- new_tibble(list(a = 5, a = 6), nrow = 1)
+
+# Old:
+as_tibble(m, validate = TRUE)
+as_tibble(df, validate = FALSE)
+
+# New:
+as_tibble(m, .name_repair = "unique")
+as_tibble(df, .name_repair = "minimal")
+```
+
+The existing `tidy_names()` and `set_tidy_names()` functions are soft-deprecated but still available unchanged.
+New code should use `.name_repair`.
+
+```{r}
+df <- new_tibble(list(a = 5, a = 6), nrow = 1)
+
+# Old:
+tidy_names(names(df))
+set_tidy_names(df)
+
+# New:
+names(as_tibble(
+  rlang::rep_named(names(df), list(NA)),
+  .name_repair = "universal"
+))
+as_tibble(df, .name_repair = "universal")
+```
+
+Assigning invalid names via `names<-()` seems like a bad idea in general, and now warns once per session.
```

Comment:
I don't think we should mention this at all, unless there is significant code that does so.

Reply:
In 3000 packages, chances are that one of them activates this. Maybe if we leave out the circumvention?